src/main/java/org/rumbledb/compiler/VariableDependenciesVisitor.java

```diff
@@ -0,0 +1,466 @@
+import org.rumbledb.expressions.module.VariableDeclaration;
+import org.rumbledb.expressions.postfix.DynamicFunctionCallExpression;
+import org.rumbledb.expressions.postfix.PredicateExpression;
+import org.rumbledb.expressions.primary.ContextItemExpression;
+import org.rumbledb.expressions.primary.FunctionCallExpression;
+import org.rumbledb.expressions.primary.InlineFunctionExpression;
+import org.rumbledb.expressions.primary.NamedFunctionReferenceExpression;
+import org.rumbledb.expressions.primary.VariableReferenceExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpression;
+import org.rumbledb.expressions.quantifiers.QuantifiedExpressionVar;
+
+
+/**
+ * Dynamic context visitor. Populates the dynamic context to evaluate the main expression.
+ */
+public class VariableDependenciesVisitor extends AbstractNodeVisitor<Void> {
+
+    private RumbleRuntimeConfiguration rumbleRuntimeConfiguration;
+
+    Map<Node, Set<String>> inputVariableDependencies;
+    Map<Node, Set<String>> outputVariableDependencies;
+
+    private void addInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            this.inputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.inputVariableDependencies.get(node).addAll(variables);
+    }
+
+    private void removeInputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            return;
+        }
+        for (String v : variables) {
+            this.inputVariableDependencies.get(node).remove(v);
+        }
+    }
+
+    private void addInputVariableDependency(Node node, String variable) {
+        if (variable == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            this.inputVariableDependencies.put(node, new TreeSet<String>());
+        }
+        this.inputVariableDependencies.get(node).add(variable);
+    }
+
+    private void removeInputVariableDependency(Node node, String variable) {
+        if (variable == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
+            return;
+        }
+        if (!this.inputVariableDependencies.keySet().contains(node)) {
+            return;
+        }
+        this.inputVariableDependencies.get(node).remove(variable);
+    }
+
+    private void addOutputVariableDependencies(Node node, Set<String> variables) {
+        if (variables == null) {
+            addInputVariableDependencies(node, Collections.emptySet());
```

Comment:
Shouldn't this be outputVariableDependencies?

Reply:
Yes, but this is now gone because of the getters.